modules/fedimint-walletv2-server/src/lib.rs

```diff
@@ -0,0 +1,1156 @@
+        dbtx: &mut DatabaseTransaction<'_>,
+    ) -> Vec<WalletConsensusItem> {
+        let mut items = dbtx
+            .find_by_prefix(&UnsignedTransactionPrefix)
+            .await
+            .map(|(key, unsigned_transaction)| {
+                let signatures = self.sign_transaction(&unsigned_transaction);
+
+                assert!(
+                    self.verify_signatures(
+                        &unsigned_transaction,
+                        &signatures,
+                        self.cfg.private.bitcoin_sk.public_key(secp256k1::SECP256K1)
+                    )
+                    .is_ok(),
+                    "Our signatures failed verification against our private key"
+                );
+
+                WalletConsensusItem::Signatures(key.0, signatures)
+            })
+            .collect::<Vec<WalletConsensusItem>>()
+            .await;
+
+        if let Ok(block_count) = self.btc_rpc.get_block_count().await {
+            items.push(WalletConsensusItem::BlockCount(
+                block_count.saturating_sub(self.cfg.consensus.finality_delay),
+            ));
+        }
+
+        // If we cannot fetch a up to date feerate from our bitcoin backend we need to
+        // retract our last vote on the feerate as an outdated consensus feerate can get
+        // our transactions stuck.
+        if let Ok(Some(fee_rate)) = self.btc_rpc.get_fee_rate(CONFIRMATION_TARGET).await {
+            items.push(WalletConsensusItem::Feerate(Some(fee_rate.sats_per_kvb)));
+        } else {
+            // Regtest node never returns fee rate
+            if self.cfg.consensus.network == Network::Regtest {
+                items.push(WalletConsensusItem::Feerate(Some(1000)));
+            } else {
+                items.push(WalletConsensusItem::Feerate(None));
+            }
+        }
+
+        items
+    }
+
+    async fn process_consensus_item<'a, 'b>(
+        &'a self,
+        dbtx: &mut DatabaseTransaction<'b>,
+        consensus_item: WalletConsensusItem,
+        peer: PeerId,
+    ) -> anyhow::Result<()> {
+        match consensus_item {
+            WalletConsensusItem::BlockCount(block_count_vote) => {
+                let current_vote = dbtx.get_value(&BlockCountVoteKey(peer)).await.unwrap_or(0);
+
+                ensure!(
+                    block_count_vote > current_vote,
+                    "Block count vote is redundant"
+                );
+
+                let old_consensus_block_count = self.consensus_block_count(dbtx).await;
+
+                dbtx.insert_entry(&BlockCountVoteKey(peer), &block_count_vote)
+                    .await;
+
+                let new_consensus_block_count = self.consensus_block_count(dbtx).await;
+
+                assert!(old_consensus_block_count <= new_consensus_block_count);
+
+                // We do not sync blocks that predate the federation itself.
+                if old_consensus_block_count == 0 {
+                    return Ok(());
+                }
```

Comment:
I'm trying to wrap my head around why this is safe. I don't see anything preventing an on-chain receive from occurring before a consensus block is reached. And I see that in some of the tests you put this comment:

```
    // We need the consensus block count to reach a non-zero value before we send in
    // any funds such that the UTXO is tracked by the federation.
```

Are we assuming that the guardians will reach block consensus before inviting anyone to their federation? That seems like a reasonable assumption to make, but we should make that requirement clear.